### ISSUE TITLE
Ingress returns 404 when `host` is set

### ISSUE THREAD

**Issue by yellowhat** [cilium/cilium#26409]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Hi,
I have installed cilium `v1.14.0-snapshot.4` to take advantage of the [l2-announcements](https://docs.cilium.io/en/latest/network/l2-announcements) feature.
As you can see from my `values.yaml` below, I have also enabled the `ingressController` feature.

If I apply the following manifest:

<details>
  <summary>Click me</summary>
  
```yaml
---
kind: Pod
apiVersion: v1
metadata:
  name: apple-app
  labels:
    app: apple
spec:
  containers:
    - name: apple-app
      image: hashicorp/http-echo
      args:
        - "-text=apple"

---
kind: Service
apiVersion: v1
metadata:
  name: apple-service
spec:
  selector:
    app: apple
  ports:
    - port: 5678 # Default port for image

---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-ingress
spec:
  ingressClassName: cilium
  rules:
    - host: <hostname>
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: apple-service
                port:
                  number: 5678
````
</details>

An ip is correctly assigned:

```console
$ kubectl get ingress
NAME              CLASS    HOSTS             ADDRESS           PORTS   AGE
example-ingress   cilium   <hostname>   192.168.178.205   80      33s
```

If I try to access:

```console
$ curl -v 192.168.178.205
*   Trying 192.168.178.205:80...
* Connected to 192.168.178.205 (192.168.178.205) port 80 (#0)
> GET / HTTP/1.1
> Host: 192.168.178.205
> User-Agent: curl/8.1.2
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< date: Wed, 21 Jun 2023 19:08:04 GMT
< server: envoy
< content-length: 0
< 
* Connection #0 to host 192.168.178.205 left intact
```

Instead I add the `Host` header, it works:

```console 
$ curl -H "Host: <hostname>" -v 192.168.178.205
*   Trying 192.168.178.205:80...
* Connected to 192.168.178.205 (192.168.178.205) port 80 (#0)
> GET / HTTP/1.1
> Host: test.mukka.haus
> User-Agent: curl/8.1.2
> Accept: */*
> 
< HTTP/1.1 200 OK
< x-app-name: http-echo
< x-app-version: 0.2.3
< date: Wed, 21 Jun 2023 19:08:39 GMT
< content-length: 6
< content-type: text/plain; charset=utf-8
< x-envoy-upstream-service-time: 2
< server: envoy
< 
apple
* Connection #0 to host 192.168.178.205 left intact
```

As you can see I have always pass the `Host:<hostname>` header.

If I remove the `host` option from the ingress object:

```yaml
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-ingress
spec:
  ingressClassName: cilium
  rules:
    - http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: apple-service
                port:
                  number: 5678
```

I can access without problem:

```console
$ curl  -v 192.168.178.205
*   Trying 192.168.178.205:80...
* Connected to 192.168.178.205 (192.168.178.205) port 80 (#0)
> GET / HTTP/1.1
> Host: 192.168.178.205
> User-Agent: curl/8.1.2
> Accept: */*
> 
< HTTP/1.1 200 OK
< x-app-name: http-echo
< x-app-version: 0.2.3
< date: Wed, 21 Jun 2023 19:10:36 GMT
< content-length: 6
< content-type: text/plain; charset=utf-8
< x-envoy-upstream-service-time: 0
< server: envoy
< 
apple
* Connection #0 to host 192.168.178.205 left intact
```

Is there something I need  to configure for the cilium ingress controller?

Thanks

### Cilium Version

```yaml
  - name: cilium
    version: v1.14.0-snapshot.4
    repository: https://helm.cilium.io
```

```yaml
---
cilium:
  # Required to enable `ingressController`
  kubeProxyReplacement: strict
  k8sServiceHost: 192.168.178.31
  k8sServicePort: 6443

  devices: enp2s0

  l2announcements:
    enabled: true
    leaseDuration: 15s
    leaseRenewDeadline: 5s
    leaseRetryPeriod: 2s

  ingressController:
    enabled: true
    enforceHttps: false
    loadbalancerMode: dedicated

  operator:
    enabled: true
    replicas: 1
    prometheus:
      enabled: true
      serviceMonitor:
        enabled: true
        interval: 30s
        # labels:
        #   release: system

  prometheus:
    enabled: true
    serviceMonitor:
      enabled: true
      interval: 30s
      # labels:
      #   release: system

  hubble:
    enabled: false
```

```yaml
---
apiVersion: cilium.io/v2alpha1
kind: CiliumLoadBalancerIPPool
metadata:
  name: default
spec:
  cidrs:
    # 192.168.178.200|220
    - cidr: 192.168.178.200/29
    - cidr: 192.168.178.208/29
    - cidr: 192.168.178.216/30
    - cidr: 192.168.178.220/32
```

```yaml
---
apiVersion: cilium.io/v2alpha1
kind: CiliumL2AnnouncementPolicy
metadata:
  name: default
spec:
  interfaces:
    - enp2s0
  externalIPs: true
  loadBalancerIPs: true
```

### Kernel Version

```console
Linux mule 6.3.8-200.fc38.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Jun 15 02:15:40 UTC 2023 x86_64 GNU/Linux
```

### Kubernetes Version

```console
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:53:42Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:47:40Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"linux/amd64"}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by margamanterola** [comment]
I believe this is just the expected behavior of the ingress resource as documented in the Kubernetes reference: https://kubernetes.io/docs/concepts/services-networking/ingress/#name-based-virtual-hosting

> If you create an Ingress resource without any hosts defined in the rules, then any web traffic to the IP address of your Ingress controller can be matched without a name based virtual host being required.

**Comment by yellowhat** [comment]
Sorry for reopening.

Some time ago I remember that even if the `- host: <hostname>` key was set, I did not need to pass the `Host: <hostname>` header.
This was on GCP.

That was the main reason to open this issue in the first place.

Is there something special done on GCP or I am misremembering?

Thanks

**Comment by margamanterola** [comment]
:woman_shrugging: If it wasn't needed, it would have been a bug, because this is the documented and expected behavior.

**Comment by bernardgut** [comment]
Hello @margamanterola  sorry to resurrect this but this breaks a lot of apps for example I migrated some fastapi backends to a cluster managed by cilium and now while you can access `/` the backend scripts will break and will display things like:

![image](https://github.com/user-attachments/assets/ca052509-7050-4f1d-99a8-94947ef0b11e)

I observe the same thing with VueJS ReactDom server. Unless I am missing somthing I think this is due to this Cilium limitation.

@margamanterola I believe this is a behavior that we should be able to change through cilium configuration settings as my backends are hostname-agnostic. In fact the whole point of using kubernetes is to have hostname-agnostic pods. Why would I want to hardcode the ingress hostnames within pods at the application level ? This doesn't make sense.

**Comment by margamanterola** [comment]
@bernardgut If it is indeed the same issue as this one, then you should be able to solve it by removing the host of the ingress definition, as defined in the Kubernetes documentation.

If that doesn't solve it, then please open a new issue, including your Ingress configuration, the paths that work and the paths that don't work, and the rest of the data required of the issue template.

**Comment by bernardgut** [comment]
Actually I dug in a bit and while reading the headers in chrome for another app I saw that the `HOST` value was actually always set properly in the headers regardless of the app config: GREAT! So then I wondered. What is the issue... After digging a whole lot more I saw that changing 

```pathType: ImplementationSpecific```

to

```
pathType: Prefix
```

fixes all the 404 issues for all apps. 

Not sure what is up with `pathType: ImplementationSpecific`. I tried to search the issues and found this : https://github.com/cilium/cilium/issues/28852. Then I checked my certificates post migration to Cilium, and, true enough I see :
```
  status:
    presented: true
    processing: true
    reason: 'Waiting for HTTP-01 challenge propagation: wrong status code ''404'',
      expected ''200'''
    state: pending

```

So yeah. Something is up with `pathType: ImplementationSpecific`. It breaks applications scripts and it also breaks letsencrypt. Sorry to necro this thread I genuinely thought it was a header `host=` issue.
